Basic/2DSpaceShooter/Assets/Scripts/Bullet.cs

```diff
@@ -46,27 +45,26 @@ private void DestroyBullet()
     
     void OnCollisionEnter2D(Collision2D other)
     {
+        var otherObject = other.gameObject;
+        
         if (!NetworkManager.Singleton.IsServer || !NetworkObject.IsSpawned)
         {
             return;
         }
-
-        var asteroid = other.gameObject.GetComponent<Asteroid>();
-        if (asteroid != null)
+        
+        if (otherObject.TryGetComponent<Asteroid>(out var asteroid))
         {
             asteroid.Explode();
             DestroyBullet();
+            return;
         }
-
-        var wall = other.gameObject.GetComponent<Wall>();
-        var obstacle = other.gameObject.GetComponent<Obstacle>();
-        if (m_Bounce == false && (wall != null || obstacle != null))
+        
+        if (m_Bounce == false && (otherObject.CompareTag("Wall") || otherObject.CompareTag("Obstacle")))
```

Comment:
Minor thing: Could we cache those strings, to avoid any allocations?

Reply:
These will not cause allocations. They are constant strings and the compiler will apply string interning for them.

Reply:
Great!